asteroid/complex_nn.py

```diff
@@ -0,0 +1,158 @@
+dtypes complex64/complex128, or tensors for which `.is_complex()` returns True.
+
+Note that Asteroid code has two other representations of complex numbers:
+
+- Torchaudio representation [..., 2] where [..., 0] and [..., 1] are real and
+  imaginary components, respectively
+- Asteroid style representation [..., 2 * n, ...] where real and imaginary components
+  are alternated, i.e. tensor([r1, c1, r2, c2, ...]). See `asteroid.filterbanks.transforms`
+  for details.
+"""
+import functools
+import torch
+import torchaudio
+import torch.nn.functional as F
+from torch import nn
+from asteroid.filterbanks import transforms
+
+
+# Alias to denote PyTorch native complex tensor (complex64/complex128).
+# `.is_complex()` returns True on these tensors.
+ComplexTensor = torch.Tensor
+
+
+def is_torch_complex(x):
+    return x.is_complex()
+
+
+def torch_complex_from_magphase(mag, phase):
+    return as_torch_complex((mag * torch.cos(phase), mag * torch.sin(phase)))
+
+
+def as_torch_complex(x, asteroid_dim=-2):
+    """Convert complex `x` to complex. Input may be one of:
+
+    - PyTorch native complex
+    - Torchaudio style complex
+    - Asteroid style complex
+    - Tuple or list of (real, imaginary) components
+
+    Args:
+        asteroid_dim (int, optional): Dimension to check for Asteroid-style complex.
+
+    :Raises:
+        ValueError: If type of `x` is not understood.
+    """
+    if isinstance(x, (list, tuple)) and len(x) == 2:
+        return torch.view_as_complex(torch.stack(x, dim=-1))
+    elif is_torch_complex(x):
+        return x
+    else:
+        is_torchaudio_complex = transforms.is_torchaudio_complex(x)
+        is_asteroid_complex = transforms.is_asteroid_complex(x, asteroid_dim)
+        if is_torchaudio_complex and is_asteroid_complex:
+            raise ValueError(
+                f"Tensor of shape {x.shape} is both a valid Torchaudio-style and "
+                "Asteroid-style complex. PyTorch complex conversion is ambiguous."
+            )
+        elif is_torchaudio_complex:
+            return torch.view_as_complex(x)
+        elif is_asteroid_complex:
+            return torch.view_as_complex(transforms.to_torchaudio(x, asteroid_dim))
+        else:
+            raise ValueError(
+                f"Do not know how to convert tensor of shape {x.shape}, dtype={x.dtype} to complex"
+            )
+
+
+def on_components(f):
+    """Make a complex-valued function callable from a real-valued one by applying it to
+    the real and imaginary components independently.
+
+    :Return:
+        cf(x), complex version of `f`: A function that applies `f` to the real and
+        imaginary components of `x` and returns the result as PyTorch complex tensor.
+    """
+
+    @functools.wraps(f)
+    def cf(x):
+        return as_torch_complex((f(x.real), f(x.imag)))
+
+    cf.__name__ == f"{f.__name__} (complex)"
+    cf.__qualname__ == f"{f.__qualname__} (complex)"
+    return cf
+
+
+class OnComponents(nn.Module):
```

Comment:
It is different: Your code uses only a single instance of `m` and thus only 1x the weights

Reply:
It can perform (documented) copy of the instance right? 
But maybe you're right that the from the first way, it's clearer that two modules are instantiated.